Construct refactoring

markrwilliams opened the pull request:
This introduces `construct` helpers ("macros" in `construct`'s jargon) that map to fundamental types and concepts in the TLS wire protocol. 

The helpers:

1. `PrefixedBytes` - Like [`PascalString`](http://construct.readthedocs.io/en/latest/api/string.html#construct.PascalString) except that its interface insists on `bytes`.  There is no `encoding` parameter and it raises a more informative error message when encoding something other than `bytes`.  This enables `tls` constructs to meaningfully require bytes.
2. `TLSPrefixedArray` - The [TLS vector type](https://tools.ietf.org/html/rfc5246#section-4.3).  It specializes on another `construct` and then encodes or decodes an arbitrarily long list or array of those `construct`s, prepending or reading a leading 16 bit length.  It's named after [`construct.macros.PrefixedArray`](always prepends a `UBInt16` length field).  This should help address #34, especially when used with `EnumSwitch` (described in 4 below).
3. `EnumClass` - Given an [`enum.Enum`](https://docs.python.org/3/library/enum.html#enum.Enum) and a `construct`, this maps that enum to that construct.  This requires some discussion as it moves enum member selection into the construct layer, and thus runs into the same problem as described in #72.  It'll probably also benefit from the same solution.  (The name isn't great, but I wanted to distinguish it from `EnumSwitch` as it just takes an `Enum` subclass)
4. `EnumSwitch` - This takes an [`enum.Enum`](https://docs.python.org/3/library/enum.html#enum.Enum) and a `construct` to represent it, then uses a dictionary mapping that enum's members to arbitrary subconstructs to translate between a member and a subconstruct.  It's intended to aid in implementing [TLS variants](https://tools.ietf.org/html/rfc5246#section-4.6.1) (i.e., `select (E) { .. }`).  It's not currently used in this PR's code because the code organization issue identified in #72 would have resulted in too large a diff; however, I included it hoping to get some feedback.  My intent is to refine this and then combine it with #78 to close #31.

The PR also makes use of the first 3 helpers in places that required minimal code reorganization.

Thanks for taking the time!

ashfall commented on tls/_constructs.py:
```diff
@@ -19,24 +20,29 @@
-    UBInt16("length"),  # TODO: Reject packets with length > 2 ** 14
-    Bytes("fragment", lambda ctx: ctx.length),
+    PrefixedBytes("fragment", UBInt16("length")),  # TODO: Reject
+                                                   # packets with
```
Mildly offtopic note - we probably should start thinking about the TODO marked here before we get too far in the implementation to come back and address this.

ashfall commented on tls/_constructs.py:
```diff
@@ -19,24 +20,29 @@
+    PrefixedBytes("fragment", UBInt16("length")),  # TODO: Reject
+                                                   # packets with
+                                                   # length > 2 ** 14
+                                                   # + 1024
```
Also, how do you feel about putting the comment on its own line (33 here, probably + 24 and 43)

ashfall commented on tls/utils.py:
```diff
@@ -24,3 +26,44 @@ def _decode(self, obj, context):
+        return obj
+
+
+def PrefixedBytes(name, length_field=construct.UBInt8("length")):  # noqa
```
Can you add a docstring here, describing this, and noting the default length field?

ashfall commented on tls/utils.py:
```diff
@@ -24,3 +26,44 @@ def _decode(self, obj, context):
+    )
+
+
+def TLSPrefixedArray(subconn, length_name="length"):  # noqa
```
Same as above, can you please add a docstring? The description you wrote in the PR sounds like it could do well as a docstring. :)

ashfall commented on tls/utils.py:
```diff
@@ -24,3 +26,44 @@ def _decode(self, obj, context):
+        length_field=construct.UBInt16(length_name))
+
+
+def EnumClass(type_field, type_enum):  # noqa
```
Docstring.

ashfall commented on tls/utils.py:
```diff
@@ -24,3 +26,44 @@ def _decode(self, obj, context):
+    return construct.SymmetricMapping(type_field, mapping)
+
+
+def EnumSwitch(type_field, type_enum, value_field, value_choices):  # noqa
```
Docstring here.

ashfall commented on tls/_constructs.py:
```diff
@@ -66,16 +66,16 @@
     SessionID,
-    CipherSuites,
+    # TODO: reject hellos with cipher_suites of length 0
+    TLSPrefixedArray(EnumClass(UBInt8("cipher_suites"), CipherSuites)),
```
This is nice, thanks!

ashfall commented on tls/test/test_utils.py:
```diff
@@ -28,3 +33,160 @@ def test_decode(self, byte, number):
+
+
+class TestBytesAdapter(object):
+
```
We haven't been as strict about this as we should have been, but I really think we should document our tests better. More importantly so _now_, because these are now heading fast towards being less straight-forward (and understandably so!). 

While these tests are awesome (and thanks a lot for adding them! :)), could you please add docstrings to document them? I find https://jml.io/pages/test-docstrings.html to be a handy reference when writing docstrings :)

ashfall commented:
@markrwilliams: This is a nice change, thank you for putting this together!

I left some comments inline, and will take another look once those are addressed. You might also want to rebase once #80 is merged.

Thanks again!

markrwilliams commented:
@ashfall: Thanks for the review!

I've rebased on master now that #80's merged.  Looks like it helped the build pass!

The lastest commit adds docstrings that (attempt) to follow [pyOpenSSL](https://github.com/pyca/pyopenssl)'s example.  The interlinking to Construct *should* work once we add its inventory to Sphinx's configuration.

I also renamed `IntEnum` (a test fixture) to `IntegerEnum`, because `enum` already claimed the name!

Please let me know what else this PR needs.  Thanks!!

ashfall commented on tls/utils.py:
```diff
@@ -23,4 +25,140 @@ def _decode(self, obj, context):
 
 
 def UBInt24(name):  # noqa
+    """
```
Thank you kindly!

ashfall commented:
Thank you!
